qutip/solver/options.py

```diff
@@ -0,0 +1,258 @@
+class SolverOdeOptions:
+    """
+    Class of options for evolution solvers such as :func:`qutip.mesolve` and
+    :func:`qutip.mcsolve`. Options can be specified either as arguments to the
+    SolverOptions constructor::
+
+        opts = SolverOptions(method=bdf, ...)
+
+    or by changing the class attributes after creation::
+
+        opts = SolverOptions()
+        opts.ode['method'] = 'bdf'
+
+    Returns options class to be used as options in evolution solvers.
+
+    The default can be changed by::
+
+        qutip.settings.solver.ode['method'] = 'bdf'
+
+    Options
+    -------
+    method : str {'adams', 'bdf', 'dop853', 'lsoda', 'vern7', 'vern9', 'diag'}
+        Integration method.
+
+    atol : float {1e-8}
+        Absolute tolerance.
+
+    rtol : float {1e-6}
+        Relative tolerance.
+
+    order : int {12}
+        Order of integrator (<=12 'adams', <=5 'bdf')
+
+    nsteps : int {2500}
+        Max. number of internal steps/call.
+
+    first_step : float {0}
+        Size of initial step (0 = automatic).
+
+    min_step : float {0}
+        Minimum step size (0 = automatic).
+
+    max_step : float {0}
+        Maximum step size (0 = automatic)
+
+    tidy: bool {True}
+        tidyup Hamiltonian before calculation
+
+    Operator_data_type: :class:`qutip.data.Data`, str {"input"}
+        Data type of the system, some method can overwrite it.
+
+    State_data_type: :class:`qutip.data.Data`, str {qutip.data.Dense}
+        Data type of the state, most solver can only work with `Dense`.
+
+    feedback_normalize: bool
+        Normalize the state before passing it to coefficient when using
+        feedback.
+    """
+    options = {
+        # Integration method (default = 'adams', for stiff 'bdf')
+        "method": 'adams',
+
+        "rhs": '',
+
+        # Absolute tolerance (default = 1e-8)
+        "atol": 1e-8,
+        # Relative tolerance (default = 1e-6)
+        "rtol": 1e-6,
+        # Maximum order used by integrator (<=12 for 'adams', <=5 for 'bdf')
+        "order": 12,
+        # Max. number of internal steps/call
+        "nsteps": 2500,
+        # Size of initial step (0 = determined by solver)
+        "first_step": 0,
+        # Max step size (0 = determined by solver)
+        "max_step": 0,
+        # Minimal step size (0 = determined by solver)
+        "min_step": 0,
+        # tidyup Hamiltonian before calculation (default = True)
+        "tidy": True,
+
+        "Operator_data_type": "input",
+
+        "State_data_type": "dense",
+        # Normalize the states received in feedback_args
+        "feedback_normalize": True,
+    }
+    extra_options = set()
+
+
+@optionsclass("results", SolverOptions)
+class SolverResultsOptions:
+    """
+    Class of options for Results of evolution solvers such as
+    :func:`qutip.mesolve` and :func:`qutip.mcsolve`.
+    Options can be specified when constructing SolverOptions
+
+        opts = SolverOptions(store_final_state=True, ...)
```

Comment:
```suggestion
        opts = SolverResultOptions(store_final_state=True, ...)
```

Reply:
No, the user always create a `SolverOptions` object which contain a `SolverResultOptions`. The creation is flat, everything is in  `SolverOptions.__init__`. The nested view is on modification after creation and printing:
```
opt = SolverOptions(store_final_state=True, method='bdf')
```
is equivalent to
```
opt = SolverOptions()
opt.results["store_final_state"] = True
opt.ode["method"] = "bdf"
```
I like this interface but it still feels rough...